webapp/src/main/scala/org/allenai/common/webapp/SprayClientHelpers.scala

```diff
@@ -0,0 +1,149 @@
+  *   }
+  *
+  *   // Completes in ~ 1 second, either with a Seq of 100 Foos or spray's RequestTimeoutException.
+  *   Future.sequence(Seq.fill(100) { getFoo })
+  * }}}
+  * format: ON
+  *
+  * Example: Making slow, CPU-intensive POST requests with internal rate-limiting.
+  * format: OFF
+  * {{{
+  *   import SprayClientHelpers._
+  *
+  *   val slowConnectorSetup = getConnectorSetup(
+  *     host = "bar.com",
+  *     port = 9876,
+  *     connectionTimeout = 1.second,
+  *     requestTimeout = 10.seconds,
+  *     // Limit the number of in-flight requests at any one time to 4, to avoid overloading the
+  *     // remote service.
+  *     maxConnections = 4
+  *   )
+  *
+  *   def fooToBar(foo: Foo): Future[Bar] = {
+  *     sendRequest(Post("/bar", foo), slowConnectorSetup) { response =>
+  *       response ~> unmarshal[Bar]
+  *     }
+  *   }
+  *
+  *   // Takes longer than 10 seconds! Only 4 requests will be sent over the wire at a time, and
+  *   // the 10-second request timeout doesn't apply until a request gets sent out.
+  *   Future.sequence(Seq.fill(16) { fooToBar(someFoo) })
+  * }}}
+  * format: ON
+  */
+object SprayClientHelpers {
+  /** Send an HTTP request through spray using a dedicated `HostConnectorSetup`, and process the
+    * response using the given function. This gives you much more control over how and when your
+    * request is sent over the wire / when it times out than does use of `sendReceive`.
+    * Caveat: this function is designed to practically prevent you from needing to worry about
+    * catching `AskTimeoutExceptions` when using spray, but it's not actually possible to
+    * guarantee. If you somehow send so many requests that one is internally buffered for more than
+    * `Int.MaxValue.millis`, you'll see an `AskTimeoutException` thrown.
+    * @param request the `HttpRequest` object to send over the wire
+    * @param connectorSetup the `HostConnectorSetup` object defining connection and timeout
+    * information for your request. See `getConnectionSetup` for a method of building these
+    * connectors.
+    * @param parseResponse a function from `HttpResponse` to a generic value `T` you want to
+    * extract from your request's response
+    */
+  def sendRequest[T](
+    request: HttpRequest,
+    connectorSetup: HostConnectorSetup
+  )(parseResponse: HttpResponse => T)(implicit actorSystem: ActorSystem): Future[T] = {
+    import actorSystem.dispatcher
+
+    // Spray requires a top-level timeout for sending requests into its infrastructure even
+    // though it has internal timeouts built-in. We set this ridiculously high timeout here so we
+    // never have to deal with uninformative `AskTimeoutExceptions`.
+    implicit val askTimeout = Timeout(Int.MaxValue.millis)
+
+    IO(Http).ask((request, connectorSetup)) map {
+      case response: HttpResponse => parseResponse(response)
+    }
+  }
+
+  /** Override spray's default settings for sending requests with the given timeouts.
```

Comment:
Setting the idle timeout of the `HostConnectorSettings` to infinite means that every `Actor` spray creates will live until the actor system dies, even if it's sitting around doing nothing for hours.  The total number of actors created will be bounded by the `maxConnections` argument, though, and I believe they're created lazily, so I'm not sure how much of a practical issue this could be.